### ISSUE TITLE
MemorySanitizer does not support manual stack switching

### ISSUE THREAD

**Issue by justincady** [google/sanitizers#1232]
I am seeing unexpected behavior running MemorySanitizer with a program that performs its own stack switching. Errors are reported for which I cannot see an obvious UMR, and the report adds these messages:

```
	Origin A (origin_id 0):
	  raw origin id: 0
	  Uninitialized value was created
	    <empty stack>

	ORIGIN: invalid (0). Might be a bug in MemorySanitizer origin tracking.
	This could still be a bug in your code, too!
```

The above report is emitted _without_ origin tracking enabled. If I do enable it, I get crashes with completely unreadable stack traces.

Then I remembered...both ASAN and TSAN have APIs to handle this use case (and I have used the ASAN APIs for this same program):

https://github.com/llvm/llvm-project/blob/master/compiler-rt/include/sanitizer/tsan_interface.h#L139
https://github.com/llvm/llvm-project/blob/master/compiler-rt/include/sanitizer/common_interface_defs.h#L299

Even given my limited understanding of the MSAN implementation, it appears similar APIs would be required for MSAN to work properly in this environment. MSAN relies on thread-local storage for its processing:

https://github.com/llvm/llvm-project/blob/master/compiler-rt/lib/msan/msan_thread.cpp
https://github.com/llvm/llvm-project/blob/master/compiler-rt/lib/msan/msan_linux.cpp#L233

Am I correct that manual stack switching with MSAN is not expected to work properly today?

Could similar APIs (to invoke before stack switch and after stack switch is complete) be provided for MSAN?

Thanks!

**Comment by eugenis** [comment]
Yes, this could be the reason.
Even more than the current thread pointer I'd worry about __msan_param_tls and similar things - they store shadow for the function arguments and will be overwritten if unrelated code runs on the same OS thread.

Can't promise a quick fix though, sorry.

**Comment by justincady** [comment]
Thanks for the information.

Regarding the potential fix timeline: is it something on the roadmap for an upcoming LLVM release?

**Comment by eugenis** [comment]
Unfortunately not.

**Comment by justincady** [comment]
Ok, thanks.

But at a high level, the changes required are really about making sure this data is always correct for the currently-in-use stack:

```
$ grep -R 'THREADLOCAL' msan
msan/msan_interceptors.cpp:static THREADLOCAL int in_interceptor_scope;
msan/msan_linux.cpp:static THREADLOCAL MsanThread* msan_current_thread;
msan/msan.cpp:static THREADLOCAL int msan_expect_umr = 0;
msan/msan.cpp:static THREADLOCAL int msan_expected_umr_found = 0;
msan/msan.cpp:THREADLOCAL u64 __msan_param_tls[kMsanParamTlsSize / sizeof(u64)];
msan/msan.cpp:THREADLOCAL u32 __msan_param_origin_tls[kMsanParamTlsSize / sizeof(u32)];
msan/msan.cpp:THREADLOCAL u64 __msan_retval_tls[kMsanRetvalTlsSize / sizeof(u64)];
msan/msan.cpp:THREADLOCAL u32 __msan_retval_origin_tls;
msan/msan.cpp:ALIGNED(16) THREADLOCAL u64 __msan_va_arg_tls[kMsanParamTlsSize / sizeof(u64)];
msan/msan.cpp:THREADLOCAL u32 __msan_va_arg_origin_tls[kMsanParamTlsSize / sizeof(u32)];
msan/msan.cpp:THREADLOCAL u64 __msan_va_arg_overflow_size_tls;
msan/msan.cpp:THREADLOCAL u32 __msan_origin_tls;
msan/msan.cpp:static THREADLOCAL int is_in_symbolizer;
```

Or perhaps another way to say it is: the above needs to be updated to be correct for the new fiber upon fiber switch.

Does that seem right? Or might that not be sufficient?

**Comment by eugenis** [comment]
First of all, stack limits in the Thread object must be updated when switching stacks. You can see how ASan does it. The FakeStack stuff does not apply to MSan, so it will be quite a lot simpler. Stack limits are used to prevent the fast (FP-based) unwinder from falling off a cliff; that must be the cause of the missing stack traces.

The, among the threadlocals you list, it seems that only param/retval/va_args ones need updating. It's OK to be lazy and simply memset them to zero. Check what ScopedThreadLocalStateBackup does for asynchronous signals - it's a very similar case to fiber switching.

**Comment by justincady** [comment]
One clarification on this issue: the error report originally posted in my bug description was due to the grpc library unconditionally enabling origin tracking when built for MSAN. (E.g. it makes no distinction between MSAN with or without origin tracking.) Because of this (and the method by which origin tracking enablement is detected), any program linking with the memory-sanitized grpc will by default assume origin tracking has been enabled for the entire program.

Everything else in this discussion is accurate regarding the specific problem of MSAN and stack switching. I have posted a diff for review that addresses this issue.
